### PR TITLE
AWS Sagemaker : Add unit tests

### PR DESCRIPTION
**Pull request by akartsky**
https://github.com/kubeflow/pipelines/issues/3631
Added one simple unit test for each component

### CODE REVIEW

File: components/aws/sagemaker/batch_transform/src/batch_transform.py [hunk 1]
```diff
@@ -65,8 +64,8 @@ def main(argv=None):
   logging.info('Batch Job request submitted. Waiting for completion...')
   _utils.wait_for_transform_job(client, batch_job_name)
 
-  Path(args.output_location_file).parent.mkdir(parents=True, exist_ok=True)
-  Path(args.output_location_file).write_text(unicode(args.output_location))
+  with open('args.output_location_file', 'w') as f:
+    f.write(unicode(args.output_location))
```
Comment (RedbackThomson) [inline components/aws/sagemaker/batch_transform/src/batch_transform.py:68]: This seems like a bug fix that is unrelated to this PR?

File: components/aws/sagemaker/log [hunk 2]
```diff
@@ -0,0 +1 @@
+<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite errors="0" failures="0" hostname="a483e75eb9c2.ant.amazon.com" name="pytest" skipped="0" tests="0" time="0.005" timestamp="2020-04-24T10:31:15.855852"></testsuite></testsuites>
```
Comment (RedbackThomson) [inline components/aws/sagemaker/log:1]: Please add to a .gitignore inside our directory.

File: components/aws/sagemaker/unit_tests/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# How to build this docker file
+#
+# https://github.com/kubeflow/pipelines.git
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/Dockerfile:3]: Git clone?

File: components/aws/sagemaker/unit_tests/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+#
+# cd pipelines/components/aws
+#
+# docker build ./ -f ./sagemaker/unit_tests/Dockerfile -t amazon/unit-test-aws-sagemaker-kfp-components:20200427
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/Dockerfile:7]: nit: tag doesn't need to contain datestamp. I would recommend `amazon/unit-test-sagemaker-components` perhaps?

File: components/aws/sagemaker/unit_tests/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# Unit tests for AWS Sagemaker KFP Components
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/README.md:1]: **Please** capitalise the `M` in SageMaker for all documentation.

File: components/aws/sagemaker/unit_tests/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Unit tests for AWS Sagemaker KFP Components 
+
+These unit tests test the user input parsing part of the code.
+
+## How to run these tests
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/README.md:5]: This should tell users to just use the Docker container. Perhaps instead of putting the header inside the Dockerfile, we could move those instructions out to here?

File: components/aws/sagemaker/unit_tests/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+boto3
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/requirements.txt:1]: Can we get the Dockerfile to use this file instead of hardcoding the values inside?

File: components/aws/sagemaker/unit_tests/run_all_tests.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
+export PYTHONPATH=../
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/run_all_tests.sh:1]: This is set inside the Dockerfile, it seems unnecessary here.

File: components/aws/sagemaker/unit_tests/run_all_tests.sh [hunk 9]
```diff
@@ -0,0 +1,4 @@
...
+
+coverage run -m pytest --junitxml ./log
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/run_all_tests.sh:3]: nit: Name of the report should be more descriptive than `log`.

File: components/aws/sagemaker/unit_tests/tests/testUtils.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+def check_empty_string_values(obj):
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/tests/testUtils.py:1]: This file name does not match that of our existing files. Python is typically snake_case for file names, so this would be `test_utils.py`. I understand this may be picked up by pytest, but either it won't run the file or there should be a way to configure it to ignore this file.

File: components/aws/sagemaker/unit_tests/tests/test_batch_transform.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+from unittest.mock import patch, Mock, MagicMock
+from botocore.exceptions import ClientError
+from datetime import datetime
+
+from batch_transform.src import batch_transform
+from common import _utils
+from . import testUtils
+
+
+# TODO : Errors out if model_name doesn't contain '-'
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/tests/test_batch_transform.py:13]: Could you please create an issue in our internal tracker for this problem?

Comment (RedbackThomson) [reply components/aws/sagemaker/unit_tests/tests/test_batch_transform.py:13]: Might also be worth it to tag this as an issue in the KFP issues? Thoughts @Jeffwan ?

File: components/aws/sagemaker/unit_tests/tests/test_workteam.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+from . import testUtils
+
+
+required_args = [
+  '--region', 'us-west-2',
+  '--team_name', 'test-team',
+  '--description', 'fake team'
+]
+
+class workteamTestCase(unittest.TestCase):
```
Comment (RedbackThomson) [inline components/aws/sagemaker/unit_tests/tests/test_workteam.py:19]: nit: `WorkTeam`

File: components/aws/sagemaker/unit_tests/Dockerfile (hunk 3 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/Dockerfile:3]: done

File: components/aws/sagemaker/unit_tests/README.md (hunk 5 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/README.md:1]: done

File: components/aws/sagemaker/unit_tests/requirements.txt (hunk 7 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/requirements.txt:1]: done

File: components/aws/sagemaker/unit_tests/run_all_tests.sh (hunk 8 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/run_all_tests.sh:1]: this way we can run this on local machine too

File: components/aws/sagemaker/unit_tests/run_all_tests.sh (hunk 9 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/run_all_tests.sh:3]: done 
changed to "unit_tests_log"

File: components/aws/sagemaker/unit_tests/tests/testUtils.py (hunk 10 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/tests/testUtils.py:1]: done

File: components/aws/sagemaker/unit_tests/tests/test_batch_transform.py (hunk 11 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/tests/test_batch_transform.py:13]: we have the internal ticket already SIM: P35195598

File: components/aws/sagemaker/unit_tests/tests/test_workteam.py (hunk 12 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/tests/test_workteam.py:19]: done

File: components/aws/sagemaker/unit_tests/README.md (hunk 6 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/README.md:5]: wrote both methods

File: components/aws/sagemaker/unit_tests/Dockerfile (hunk 4 above)
Comment (akartsky) [reply components/aws/sagemaker/unit_tests/Dockerfile:7]: done

File: components/aws/sagemaker/common/_utils.py [hunk 13]
```diff
@@ -228,6 +228,14 @@ def get_image_from_job(client, job_name):
 
 
 def create_model(client, args):
+    request = create_model_request(args)
+    create_model_response = client.create_model(**request)
```
Comment (gautamkmr) [inline components/aws/sagemaker/common/_utils.py:232]: Please put it inside `try/catch` and handle the exception accordingly.

Comment (gautamkmr) [reply components/aws/sagemaker/common/_utils.py:232]: check `create_training_job` for reference.

File: components/aws/sagemaker/tests/unit_tests/README.md [hunk 14]
```diff
@@ -0,0 +1,43 @@
+# Unit tests for AWS SageMaker KFP Components 
+
+These unit tests test the user input parsing part of the code.
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/README.md:3]: Either can we add more information here or simply removing this explanation.

File: components/aws/sagemaker/common/_utils.py [hunk 15]
```diff
@@ -327,12 +330,16 @@ def create_endpoint_config(client, args):
...
+def create_endpoint_config(client, args):
+    request = create_endpoint_config_request(args)
     try:
         create_endpoint_config_response = client.create_endpoint_config(**request)
         logging.info("Endpoint configuration in SageMaker: https://{}.console.aws.amazon.com/sagemaker/home?region={}#/endpointConfig/{}"
-            .format(args['region'], args['region'], endpoint_config_name))
+            .format(args['region'], args['region'], request['EndpointConfigName']))
         logging.info("Endpoint Config Arn: " + create_endpoint_config_response['EndpointConfigArn'])
-        return endpoint_config_name
+        return request['EndpointConfigName']
```
Comment (gautamkmr) [inline components/aws/sagemaker/common/_utils.py:342]: Why this change ? Some caller might need `endpoint arn`.

File: components/aws/sagemaker/batch_transform/src/batch_transform.py (hunk 1 above)
Comment (akartsky) [reply components/aws/sagemaker/batch_transform/src/batch_transform.py]: ya ... reverted

File: components/aws/sagemaker/tests/unit_tests/README.md (hunk 14 above)
Comment (akartsky) [reply components/aws/sagemaker/tests/unit_tests/README.md:3]: removed

File: components/aws/sagemaker/common/_utils.py (hunk 13 above)
Comment (akartsky) [reply components/aws/sagemaker/common/_utils.py]: done

File: components/aws/sagemaker/common/_utils.py (hunk 15 above)
Comment (akartsky) [reply components/aws/sagemaker/common/_utils.py:344]: nothing changed 
request['EndpointConfigName'] = endpoint_config_name

File: components/aws/sagemaker/log (hunk 2 above)
Comment (RedbackThomson) [reply components/aws/sagemaker/log:1]: Please remove this file.

File: components/aws/sagemaker/tests/unit_tests/requirements.txt [hunk 16]
```diff
@@ -0,0 +1,6 @@
+boto3
```
Comment (RedbackThomson) [inline components/aws/sagemaker/tests/unit_tests/requirements.txt:1]: Pin to version numbers

Comment (akartsky) [reply components/aws/sagemaker/tests/unit_tests/requirements.txt]: done

File: components/aws/sagemaker/tests/unit_tests/tests/test_batch_transform.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+  '--output_location_file', 'tmp/'
+]
+
+class BatchTransformTestCase(unittest.TestCase):
+  @classmethod
+  def setUpClass(cls):
+    parser = batch_transform.create_parser()
+    cls.parser = parser
+
+  def test_sample(self):
```
Comment (mbaijal) [inline components/aws/sagemaker/tests/unit_tests/tests/test_batch_transform.py:31]: nit: method name could be more descriptive
